src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs

```diff
@@ -1,332 +1,123 @@
+        internal override SyntaxNode GetIfStatement(TextSpan textSpan, SyntaxToken token, CancellationToken cancellationToken)
         {
-            var document = context.Document;
-            var textSpan = context.Span;
-            var cancellationToken = context.CancellationToken;
-
-            if (!textSpan.IsEmpty)
-            {
-                return;
-            }
-
-            if (document.Project.Solution.Workspace.Kind == WorkspaceKind.MiscellaneousFiles)
-            {
-                return;
-            }
-
-            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
-            var ifStatement = root.FindToken(textSpan.Start).GetAncestor<IfStatementSyntax>();
+            var ifStatement = token.GetAncestor<IfStatementSyntax>();
             if (ifStatement == null || ifStatement.Else == null)
             {
-                return;
-            }
-
-            if (!ifStatement.IfKeyword.Span.IntersectsWith(textSpan.Start))
-            {
-                return;
+                return null;
             }
 
-            if (ifStatement.OverlapsHiddenPosition(cancellationToken))
+            var span = TextSpan.FromBounds(ifStatement.GetFirstToken().Span.Start, ifStatement.CloseParenToken.Span.End);
+            if (!span.IntersectsWith(textSpan))
             {
-                return;
+                return null;
             }
 
-            context.RegisterRefactoring(
-                new MyCodeAction(
-                    CSharpFeaturesResources.Invert_if_statement,
-                    c => InvertIfAsync(document, ifStatement, c)));
+            return ifStatement;
         }
 
-        private async Task<Document> InvertIfAsync(Document document, IfStatementSyntax ifStatement, CancellationToken cancellationToken)
+        internal override SyntaxNode GetRootWithInvertIfStatement(
+            Document document,
+            SemanticModel model,
+            SyntaxNode ifStatement,
+            CancellationToken cancellationToken)
         {
-            var tree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
-            var ifNode = ifStatement;
-
-            // In the case that the else clause is actually an else if clause, place the if
-            // statement to be moved in a new block in order to make sure that the else
-            // statement matches the right if statement after the edit.
-            var newIfNodeStatement = ifNode.Else.Statement.Kind() == SyntaxKind.IfStatement ?
-                SyntaxFactory.Block(ifNode.Else.Statement) :
-                ifNode.Else.Statement;
+            var generator = SyntaxGenerator.GetGenerator(document);
+            var syntaxFacts = GetSyntaxFactsService();
 
-            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var ifNode = (IfStatementSyntax)ifStatement;
```

Comment:
ifStatement and ifNode are not good names.  perhaps oldIfStatement and newIfStatement?